File: frontend/src/types/diary.ts
// 日记相关类型定义

export type EmotionTag = string
export type EventType = 'work' | 'relationship' | 'health' | 'achievement' | 'other'

export interface Diary {
  id: number
  user_id: number
  title: string
  content: string
  content_html?: string
  diary_date: string
  emotion_tags: EmotionTag[]
  importance_score: number
  word_count: number
  media_urls: string[]
  created_at: string
  updated_at: string
  is_analyzed: boolean
}

export interface DiaryCreate {
  title: string
  content: string
  content_html?: string
  diary_date?: string
  emotion_tags?: EmotionTag[]
  importance_score?: number
  media_urls?: string[]
}

export interface DiaryUpdate {
  title?: string
  content?: string
  content_html?: string
  diary_date?: string
  emotion_tags?: EmotionTag[]
  importance_score?: number
  media_urls?: string[]
}

export interface DiaryListResponse {
  items: Diary[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface TimelineEvent {
  id: number
  diary_id: number
  event_summary: string
  emotion_tag: string
  importance_score: number
  event_type: EventType
  related_entities: Record<string, any>
  event_date: string
  created_at: string
}

export interface EmotionStats {
  tag: string
  count: number
  percentage: number
}

export interface TerrainEvent {
  id: number | null
  diary_id: number
  summary: string
  emotion_tag: string | null
  importance_score: number
  event_type: EventType | null
  source_label?: string
}

export interface TerrainPoint {
  date: string
  energy: number | null
  valence: number | null
  density: number
  events: TerrainEvent[]
}

export interface TerrainPeak {
  date: string
  value: number
  label: string
  summary: string
}

export interface TerrainValley {
  date_range: [string, string]
  min_value: number
  days: number
  label: string
  summary: string
}

export interface TerrainInsights {
  peaks: TerrainPeak[]
  valleys: TerrainValley[]
  trend: 'ascending' | 'descending' | 'stable' | 'overall' | string
  trend_description: string
}

export interface TerrainMeta {
  start_date: string
  end_date: string
  total_events: number
  days_with_data: number
  total_days: number
}

export interface TerrainResponse {
  points: TerrainPoint[]
  insights: TerrainInsights
  meta: TerrainMeta
}

export interface DashboardInsightEmotion {
  tag: string
  label: string
  count: number
  percentage: number
}

export interface DashboardInsightDiary {
  id: number
  title: string
  diary_date: string
  emotion_tags: EmotionTag[]
  summary: string
  word_count: number
  importance_score: number
  is_analyzed: boolean
  analysis_path: string
}

export interface DashboardInsights {
  window_days: number
  generated_at: string
  stats: {
    total_diaries: number
    last_days_count: number
    this_month_count: number
    top_emotion: string
    top_emotion_label: string
    top_emotion_count: number
    trend: 'ascending' | 'descending' | 'stable' | string
    trend_label: string
    trend_delta: number
    risk_label: string
    risk_desc: string
    negative_ratio: number
  }
  ai_observation: {
    title: string
    summary: string
    encouragement: string
    source: string
  }
  emotion_stats: DashboardInsightEmotion[]
  insights: string[]
  recent_diaries: DashboardInsightDiary[]
  analysis_entry: {
    overall_path: string
    single_diary_path_template: string
    description: string
  }
}

export interface GrowthDailyInsight {
  date: string
  primary_emotion?: string
  summary?: string
  has_content: boolean
  cached: boolean
  source?: string
  message?: string
}

File: backend/app/schemas/diary.py
"""
日记相关的 Pydantic Schemas
"""
from typing import Optional, List
from datetime import datetime, date
from pydantic import BaseModel, Field, field_validator


class DiaryCreate(BaseModel):
    """创建日记请求"""
    title: Optional[str] = Field(None, max_length=200, description="日记标题")
    content: str = Field(..., min_length=1, max_length=10000, description="日记内容")
    content_html: Optional[str] = Field(None, max_length=50000, description="日记HTML内容")
    diary_date: Optional[date] = Field(None, description="日记日期")
    emotion_tags: Optional[List[str]] = Field(None, description="情绪标签列表")
    importance_score: int = Field(default=5, ge=1, le=10, description="重要性评分（1-10）")
    images: Optional[List[str]] = Field(None, description="图片URL列表")

    @field_validator('diary_date', mode='before')
    @classmethod
    def set_default_date(cls, v):
        """如果没有提供日期，使用今天"""
        if v is None:
            return date.today()
        return v

    @field_validator('content')
    @classmethod
    def validate_content(cls, v):
        """验证内容不为空"""
        if not v or not v.strip():
            raise ValueError('日记内容不能为空')
        return v


class DiaryUpdate(BaseModel):
    """更新日记请求"""
    title: Optional[str] = Field(None, max_length=200, description="日记标题")
    content: Optional[str] = Field(None, min_length=1, max_length=10000, description="日记内容")
    content_html: Optional[str] = Field(None, max_length=50000, description="日记HTML内容")
    diary_date: Optional[date] = Field(None, description="日记日期")
    emotion_tags: Optional[List[str]] = Field(None, description="情绪标签列表")
    importance_score: Optional[int] = Field(None, ge=1, le=10, description="重要性评分（1-10）")
    images: Optional[List[str]] = Field(None, description="图片URL列表")
    is_analyzed: Optional[bool] = Field(None, description="是否已分析")


class DiaryResponse(BaseModel):
    """日记响应"""
    id: int
    user_id: int
    title: Optional[str]
    content: str
    content_html: Optional[str]
    diary_date: date
    emotion_tags: Optional[List[str]]
    importance_score: int
    word_count: int
    images: Optional[List[str]]
    is_analyzed: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class DiaryListResponse(BaseModel):
    """日记列表响应"""
    items: List[DiaryResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class TimelineEventCreate(BaseModel):
    """创建时间轴事件请求"""
    diary_id: Optional[int] = Field(None, description="关联的日记ID")
    event_date: date = Field(..., description="事件日期")
    event_summary: str = Field(..., min_length=1, max_length=500, description="事件摘要")
    emotion_tag: Optional[str] = Field(None, max_length=50, description="情绪标签")
    importance_score: int = Field(default=5, ge=1, le=10, description="重要性评分")
    event_type: Optional[str] = Field(None, max_length=50, description="事件类型")
    related_entities: Optional[dict] = Field(None, description="相关实体")


class TimelineEventResponse(BaseModel):
    """时间轴事件响应"""
    id: int
    user_id: int
    diary_id: Optional[int]
    event_date: date
    event_summary: str
    emotion_tag: Optional[str]
    importance_score: int
    event_type: Optional[str]
    related_entities: Optional[dict]
    created_at: datetime

    class Config:
        from_attributes = True
